api/storage/deployment_storage.go

```diff
@@ -27,6 +29,10 @@ type DeploymentStorage interface {
 	ListInModelVersion(modelID, versionID, endpointUUID string) ([]*models.Deployment, error)
 	// Save save the deployment to underlying storage
 	Save(deployment *models.Deployment) (*models.Deployment, error)
+	// OnDeploymentSuccess updates the new deployment status to successful on DB and update all previous deployment status for that version endpoint to terminated.
+	OnDeploymentSuccess(newDeployment *models.Deployment) error
```

Comment:
I'm not sure these 2 methods should be in the storage layer. The storage layer is typically slim, providing just the data access capabilities to the DB, right? Instead of these, can we introduce:

```golang
ListDeploymentsWithStatus(modelID string, versionID string, endpointUUID string, statuses []string)
```

This method can be called with the appropriate input params (statuses will be {"running", "serving"}) - then, their status can be updated and `Save()` called from the service layer.

Wdyt?

Reply:
The main reason I added it here is that I want to use a DB transaction because we are updating multiple data at once. Calling Save() multiple times in the service layer might introduce inconsistent data updates (one of Save() might fail which can lead to not updated data).

We also have a similar approach with [Model Endpoint Storage's Save method](https://github.com/caraml-dev/merlin/blob/main/api/storage/model_endpoint_storage.go#L79-L132) where it updates both model and version endpoints.

Reply:
Yes, my implementation which uses transaction will only call one save. I was referring to another approach where the service layer is the one that call db.Save() multiple times and without using transaction.